tests/integration/test_api.py

```diff
@@ -0,0 +1,336 @@
+    assert result.json == {
+        "count": 1,
+        "questions": [
+            {
+                "barcode": "1",
+                "type": "add-binary",
+                "value": "Seeds",
+                "question": "Does the product belong to this category?",
+                "insight_id": insight_id,
+                "insight_type": "category",
+            }
+        ],
+        "status": "found",
+    }
+
+
+def test_random_question_user_has_already_seen():
+    AnnotationVote.create(
+        insight_id=insight_id,
+        value=1,
+        device_id="device1",
+    )
+
+    cl = client()
+
+    result = cl.simulate_get("/api/v1/questions/random?device_id=device1")
+
+    assert result.status_code == 200
+    assert result.json == {"count": 0, "questions": [], "status": "no_questions"}
+
+
+def test_popular_question():
+    cl = client()
+    result = cl.simulate_get("/api/v1/questions/popular")
+
+    assert result.status_code == 200
+    assert result.json == {
+        "count": 1,
+        "questions": [
+            {
+                "barcode": "1",
+                "type": "add-binary",
+                "value": "Seeds",
+                "question": "Does the product belong to this category?",
+                "insight_id": insight_id,
+                "insight_type": "category",
+            }
+        ],
+        "status": "found",
+    }
+
+
+def test_barcode_question():
+    cl = client()
+    result = cl.simulate_get("/api/v1/questions/2")
+
+    assert result.status_code == 200
+    assert result.json == {"questions": [], "status": "no_questions"}
+
+    result = cl.simulate_get("/api/v1/questions/1")
+
+    assert result.status_code == 200
+    assert result.json == {
+        "questions": [
+            {
+                "barcode": "1",
+                "type": "add-binary",
+                "value": "Seeds",
+                "question": "Does the product belong to this category?",
+                "insight_id": insight_id,
+                "insight_type": "category",
+            }
+        ],
+        "status": "found",
+    }
+
+
+def test_annotate_insight_authenticated():
+    cl = client()
+    result = cl.simulate_post(
+        "/api/v1/insights/annotate",
+        params={
+            "insight_id": insight_id,
+            "annotation": -1,
+        },
+        headers={"Authorization": "Basic " + base64.b64encode(b"a:b").decode("ascii")},
+    )
+
+    assert result.status_code == 200
+    assert result.json == {"description": "the annotation was saved", "status": "saved"}
+
+    # For authenticated users we expect the insight to be validated directly, tracking the username of the annotator.
+    votes = list(AnnotationVote.select())
+    assert len(votes) == 0
+
+    insight = next(
+        ProductInsight.select()
+        .where(ProductInsight.id == insight_id)
+        .dicts()
+        .iterator()
+    )
+    assert insight.items() > {"username": "a", "annotation": -1, "n_votes": 0}.items()
```

Comment:
wow, never seen that kind of dict_items comparison before ! That's cool.

Reply:
What does it actually do? I sort of feel that it should check that it has "at least those elements", but I can't say I feel certain about it.

Reply:
@zond, yep, it does what you said :)

Reply:
Nice one, thanks!

Reply:
Yes python docs says dict_items acts as a Set of (key, value). And for sets, A > B means that A got all the elements from B and some more.